### ISSUE TITLE
Custom functions

### ISSUE THREAD

**Issue by ArrestedDevelopment** [medialize/sass.js#27]
I think it'd be awesome if there was a way to define a function like how node-sass does: (from sass/node-sass#644)

```js
sass.render({
  data: '#{headings(2,5)} { color: #08c; }',
  functions: {
    'headings($from: 0, $to: 6)': function(from, to) {
      var i, f = from.getValue(), t = to.getValue(),
          list = new sass.types.List(t - f + 1);

      for (i = f; i <= t; i++) {
        list.setValue(i - f, new sass.types.String('h' + i));
      }

      return list;
    }
  }
});
```

Right now, thanks to import hooks, some half dynamic functionality can be put into styles (like inserting the current date into a comment could be done by stuffing it into a variable in an imported "file") but custom functions could potentially allow for something like compass to be used by sass.js

**Comment by rodneyrehm** [comment]
Yes, I have custom functions on the radar. What's keeping this from happening is my inability (and lack of time to learn) to figure out the marshaling:

We need to expose [Sass_Value](https://github.com/sass/libsass/blob/3.2.0/sass_values.cpp#L72) in some sort of C++ class, so we can have emscripten marshal the data structure automatically using [WebIDL Binder](http://kripken.github.io/emscripten-site/docs/porting/connecting_cpp_and_javascript/WebIDL-Binder.html#webidl-binder). We need a two way binding (export to JS, import from JS), and for that WebIDL looks like the better option than [embind](http://kripken.github.io/emscripten-site/docs/porting/connecting_cpp_and_javascript/embind.html).

Unlike node-sass' implementation, Sass.js will have to make functions work asnychronously, which is not as big of a problem, since we already figured out how to acoomplish that using emterpreter for the Importer API.

Custom functions *will* happen, I'm just not sure when I'll get to land this. Anyone comfortable with C/C++ and emscripten can certainly have a go at it any time, though ;)

**Comment by mcampa** [comment]
Is there any progress on this issue? :)

**Comment by rodneyrehm** [comment]
Unofrtunately not. The last attempt to figure this out was in early 2016 and there's been 0 effort since. Do you want to look into this?

**Comment by mcampa** [comment]
I would love to, but I'm not familiar with C/C++ or emscripten
